<a href="https://colab.research.google.com/github/dmeadows95/CS4783_HW03/blob/main/CS4783_HW3_DamonMeadows.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Load the TensorBoard notebook extension.
%load_ext tensorboard

from datetime import datetime
from packaging import version

import tensorflow as tf
from tensorflow import keras
from keras.utils import to_categorical
from keras.models import Sequential
from keras.layers import Dense, Conv2D, Flatten, MaxPooling2D

import tensorboard
import matplotlib.pyplot as plt

import numpy as np

In [ ]:
# import data
(train_images, train_labels), (test_images, test_labels) = keras.datasets.mnist.load_data()
train_images = train_images / 255.0

# reshape data
train_images = train_images.reshape(60000,28,28,1)
test_images = test_images.reshape(10000,28,28,1)

# one-hot encode data
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

In [ ]:
# create model 1
# model 1 => number of filters increases each layer
model_1 = Sequential()
# add model layers
model_1.add(Conv2D(4, kernel_size=3, activation='relu', input_shape=(28,28,1)))
model_1.add(Conv2D(8, kernel_size=3, activation='relu'))
model_1.add(Conv2D(16, kernel_size=3, activation='relu'))
model_1.add(Conv2D(32, kernel_size=3, activation='relu'))
model_1.add(Conv2D(64, kernel_size=3, activation='relu'))
model_1.add(Conv2D(128, kernel_size=3, activation='relu'))
model_1.add(Conv2D(256, kernel_size=3, activation='relu'))
model_1.add(Conv2D(512, kernel_size=3, activation='relu'))
model_1.add(Conv2D(1024, kernel_size=3, activation='relu'))
model_1.add(Conv2D(2048, kernel_size=3, activation='relu'))

model_1.add(Flatten())
model_1.add(Dense(10, activation='softmax'))

In [ ]:
from keras import backend as K

optimizer = keras.optimizers.Adam(lr=.0001)

# compile model using accuracy to measure model performance
model_1.compile(
      optimizer=optimizer, 
      loss='categorical_crossentropy',
      metrics=['accuracy'])

model_1.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_50 (Conv2D)          (None, 26, 26, 4)         40        
                                                                 
 conv2d_51 (Conv2D)          (None, 24, 24, 8)         296       
                                                                 
 conv2d_52 (Conv2D)          (None, 22, 22, 16)        1168      
                                                                 
 conv2d_53 (Conv2D)          (None, 20, 20, 32)        4640      
                                                                 
 conv2d_54 (Conv2D)          (None, 18, 18, 64)        18496     
                                                                 
 conv2d_55 (Conv2D)          (None, 16, 16, 128)       73856     
                                                                 
 conv2d_56 (Conv2D)          (None, 14, 14, 256)      

In [ ]:
# fit model_1
model_1.fit(
    train_images, 
    train_labels, 
    batch_size = 32,
    validation_data=(test_images, test_labels), epochs=3)

Epoch 1/3
1875/1875 [==============================] - 149s 79ms/step - loss: 0.2059 - accuracy: 0.9356 - val_loss: 10.9804 - val_accuracy: 0.9736
Epoch 2/3
1875/1875 [==============================] - 146s 78ms/step - loss: 0.0682 - accuracy: 0.9791 - val_loss: 7.6123 - val_accuracy: 0.9782
Epoch 3/3
1875/1875 [==============================] - 142s 76ms/step - loss: 0.0497 - accuracy: 0.9847 - val_loss: 6.4405 - val_accuracy: 0.9805


In [ ]:
# Evaluate
score = model_1.evaluate(test_images, test_labels)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

313/313 [==============================] - 7s 21ms/step - loss: 3.1540 - accuracy: 0.1011
Test loss: 3.1539838314056396
Test accuracy: 0.10109999775886536


Model 1 Evaluation Notes:<br><br>
Batch Size Trials (SGD, alpha = .01, categorical cross entropy, epcochs =3):<br><br>
Batch size = 8:<br>
trial 1: time = 1100s; accuracy = 98.55%<br>
trial 2: time = 1082s; accuracy = 98.43%<br>
trial 3: time = 1095s; accuracy = 98.32%<br>
trial 4: time = 1082s; accuracy = 98.82%<br>
trial 5: time = 1089s; accuracy = 98.66%<br>
average time =1089.6s; average accuracy = 98.556%<br><br>
Batch size = 16:<br>
trial 1: time = 600s; accuracy = 98.53%<br>
trial 2: time = 606s; accuracy = 98.48%<br>
trial 3: time = 604s; accuracy = 98.38%<br>
trial 4: time = 609s; accuracy = 98.44%<br>
trial 5: time = 606s; accuracy = 98.56%<br>
average time = 605s; average accuracy = 98.478%<br>
Batch size = 32:<br>
trial 1: time = 407s; accuracy = 98.35%<br>
trial 2: time = 410s; accuracy = 98.26%<br>
trial 3: time = 407s; accuracy = 98.26%<br>
trial 4: time = 411s; accuracy = 98.08%<br>
trial 5: time = 407s; accuracy = 97.82%<br>
average time = 408.4s; average accuracy = 98.154%<br>
Batch size = 64:<br>
trial 1: time = 307s; accuracy = 96.60%<br>
trial 2: time = 306s; accuracy = 97.21%<br>
trial 3: time = 305s; accuracy = 98.06%<br>
trial 4: time = 305s; accuracy = 96.88%<br>
trial 5: time = 306s; accuracy = 92.27%<br>
average time = 305.8s; average accuracy = 96.204%<br>
Batch size = 128:<br>
trial 1: time = 269s; accuracy = 94.65%<br>
trial 2: time = 260s; accuracy = 96.99%<br>
trial 3: time = 262s; accuracy = 96.34%<br>
trial 4: time = 259s; accuracy = 97.48%<br>
trial 5: time = 259s; accuracy = 96.60%<br>
average time = 261.8s; average accuracy = 96.412%<br><br>

Optimizer trial (batch size = 128, alpha = default per optimizer unless specified, loss=categorical cross entropy, epochs=3):<br><br>

Optimizer = adam:<br>
trial 1: time = 247s; accuracy = 96.52%<br>
trial 2: time = 249s; accuracy = 98.00%<br>
trial 3: time = 252s; accuracy = 89.98%<br>
trial 4: time = 287s; accuracy = 95.91%<br>
trial 5: time = 263s; accuracy = 94.71%<br>
average time = 259.6s; average test accuracy = 95.024%<br>
Optimizer = adam, alpha = .01:<br>
trial 1: time = 249s; training accuracy = 11.14%; test accuracy = 12.15%<br>
Optimizer = adam, alpha = .0001:<br>
trial 1: time = 288s; training accuracy = 98.23%; test accuracy = 98.28%<br>
trial 2: time = 288s; training accuracy = 98.26%; test accuracy = 98.54%<br>
trial 3: time = 289s; training accuracy = 97.74%; test accuracy = 95.42%<br>
trial 4: time = 287s; training accuracy = 98.05%; test accuracy = 97.55%<br>
trial 5: time = 288s; training accuracy = 98.13%; test accuracy = 98.28%<br>
average time = 288s; average training accuracy = 98.082; average test accuracy = 97.614
Optimizer = adam, alpha = .00001:<br>
trial 1: time = 289s; training accuracy = 95.46%; test accuracy = 96.40%<br>
trial 2: time = 301s; training accuracy = 94.93%; test accuracy = 94.59%<br>
trial 3: time = 293s; training accuracy = 95.23%; test accuracy = 95.93%<br>
Optimizer = adam, alpha = .0005:<br>
trial 1: time = 281s; training accuracy = 98.49%; test accuracy = 97.15%<br>
trial 2: time = 279s; training accuracy = 98.55%; test accuracy = 98.05%<br>
trial 3: time = 281s; training accuracy = 98.40%; test accuracy = 95.42%<br>
trial 4: time = 278s; training accuracy = 98.43%; test accuracy = 92.48%<br>
trial 5: time = 277s; training accuracy = 98.42%; test accuracy = 98.07%<br>
average time = 279.2s; average training accuracy = 98.458%; average test accuracy = 96.2276%<br>

current best:
optimizer = adam; alpha = .0001: average test accuracy = 97.614<br><br>

Optimizer SGD, alpha = .001:<br>
trial 1: time = 288s; training accuracy = 11.30%; test accuracy = 21.66%<br>
Optimizer = SGD, alpha = .1:<br>
trial 1: time = 285s; training accuracy = 97.86%; test accuracy = 93.61%<br>
trial 2: time = 282s; training accuracy = 97.28%; test accuracy = 97.13%<br>
trial 3: time = 271s; training accuracy = 11.24%; test accuracy = 10.11%<br><br>

Potential Final:
optimizer = adam; alpha = .0001; batch size = 32; loss = categorical cross entropy; epochs = 3:<br>
trial 1: time = 433s; training accuracy = 98.50%; test accuracy = 97.70%<br>
trial 2: time = 426s; training accuracy = 98.40%; test accuracy = 96.99%<br>
trial 3: time = 437s; training accuracy = 98.31%; test accuracy = 98.35%<br>
trial 4: time = 435%; training accuracy = 98.41%; test accuracy = 97.25%<br>
trial 5: time = 437s; training accuracy = 98.47%; test accuracy = 98.05%<br>




In [ ]:
# create model 2
# model 2 => number of filters descreases each layer
model_2 = Sequential()
# add model layers
model_2.add(Conv2D(4096, kernel_size=3, activation='relu', input_shape=(28,28,1)))
model_2.add(Conv2D(2048, kernel_size=3, activation='relu'))
model_2.add(Conv2D(1024, kernel_size=3, activation='relu'))
model_2.add(Conv2D(512, kernel_size=3, activation='relu'))
model_2.add(Conv2D(256, kernel_size=3, activation='relu'))
model_2.add(Conv2D(128, kernel_size=3, activation='relu'))
model_2.add(Conv2D(64, kernel_size=3, activation='relu'))
model_2.add(Conv2D(32, kernel_size=3, activation='relu'))
model_2.add(Conv2D(16, kernel_size=3, activation='relu'))
model_2.add(Conv2D(8, kernel_size=3, activation='relu'))

model_2.add(Flatten())
model_2.add(Dense(10, activation='softmax'))

In [ ]:
optimizer = keras.optimizers.Adam(lr=.001)

# compile model using accuracy to measure model performance
model_2.compile(
      optimizer=optimizer, 
      loss='categorical_crossentropy',
      metrics=['accuracy'])

model_2.summary()

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_70 (Conv2D)          (None, 26, 26, 4096)      40960     
                                                                 
 conv2d_71 (Conv2D)          (None, 24, 24, 2048)      75499520  
                                                                 
 conv2d_72 (Conv2D)          (None, 22, 22, 1024)      18875392  
                                                                 
 conv2d_73 (Conv2D)          (None, 20, 20, 512)       4719104   
                                                                 
 conv2d_74 (Conv2D)          (None, 18, 18, 256)       1179904   
                                                                 
 conv2d_75 (Conv2D)          (None, 16, 16, 128)       295040    
                                                                 
 conv2d_76 (Conv2D)          (None, 14, 14, 64)       

In [ ]:
# fit model_2
model_2.fit(
    train_images, 
    train_labels, 
    batch_size = 256,
    validation_data=(test_images, test_labels), epochs=3)

Epoch 1/3


Model 2 Evaluation Notes:<br>

Initial runs at default optimization values:<br>
Adam, batch size = 256:<br>
at batch size 128 the testing was still over an hour, at 256 the runtime broke and I had to restart.


In [ ]:
# create model 3
# model 3 => number of filters increases first 5 layers then decreases
model_3 = Sequential()
# add model layers
model_3.add(Conv2D(32, kernel_size=3, activation='relu', input_shape=(28,28,1)))
model_3.add(Conv2D(64, kernel_size=3, activation='relu'))
model_3.add(Conv2D(128, kernel_size=3, activation='relu'))
model_3.add(Conv2D(256, kernel_size=3, activation='relu'))
model_3.add(Conv2D(512, kernel_size=3, activation='relu'))
model_3.add(Conv2D(512, kernel_size=3, activation='relu'))
model_3.add(Conv2D(256, kernel_size=3, activation='relu'))
model_3.add(Conv2D(128, kernel_size=3, activation='relu'))
model_3.add(Conv2D(64, kernel_size=3, activation='relu'))
model_3.add(Conv2D(32, kernel_size=3, activation='relu'))

model_3.add(Flatten())
model_3.add(Dense(10, activation='softmax'))

In [ ]:
optimizer = keras.optimizers.SGD(lr=.03)

# compile model using accuracy to measure model performance
model_3.compile(
      optimizer=optimizer, 
      loss='categorical_crossentropy',
      metrics=['accuracy'])

model_3.summary()

Model: "sequential_21"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_210 (Conv2D)         (None, 26, 26, 32)        320       
                                                                 
 conv2d_211 (Conv2D)         (None, 24, 24, 64)        18496     
                                                                 
 conv2d_212 (Conv2D)         (None, 22, 22, 128)       73856     
                                                                 
 conv2d_213 (Conv2D)         (None, 20, 20, 256)       295168    
                                                                 
 conv2d_214 (Conv2D)         (None, 18, 18, 512)       1180160   
                                                                 
 conv2d_215 (Conv2D)         (None, 16, 16, 512)       2359808   
                                                                 
 conv2d_216 (Conv2D)         (None, 14, 14, 256)     

In [ ]:
# fit model 3
model_3.fit(
    train_images, 
    train_labels, 
    batch_size = 16,
    validation_data=(test_images, test_labels), epochs=3)

Epoch 1/3
3750/3750 [==============================] - 93s 25ms/step - loss: 0.2896 - accuracy: 0.9074 - val_loss: 8.9418 - val_accuracy: 0.9731
Epoch 2/3
3750/3750 [==============================] - 91s 24ms/step - loss: 0.0675 - accuracy: 0.9792 - val_loss: 5.9725 - val_accuracy: 0.9796
Epoch 3/3
3750/3750 [==============================] - 91s 24ms/step - loss: 0.0480 - accuracy: 0.9854 - val_loss: 6.4989 - val_accuracy: 0.9817


In [ ]:
# Evaluate
score = model_3.evaluate(test_images, test_labels)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

313/313 [==============================] - 4s 13ms/step - loss: 7.0393 - accuracy: 0.9734
Test loss: 7.039312362670898
Test accuracy: 0.9733999967575073


Model 3 Evaluation Notes:<br>
Constant Parameters: loss = categorical cross entropy, epochs = 3<br>

Initial optimizer comparison:<br>
Adam, batch size = 64, alpha = .001(default):<br>
trial 1 = 221s; training accuracy = 98.21%; test accuracy = 94.49%<br>
trial 2 = 206s; training accuracy = 98.12%; test accuracy = 96.96%<br>
trial 3 = 217s; training accuracy = 98.22%; test accuracy = 94.72%<br>
SGD, batch size = 64, alpha = .01(default):<br>
trial 1 = 209s; training accuracy = 97.00%; test accuracy = 96.64%<br>
trial 2 = 210s; training accuracy = 96.45%; test accuracy = 96.48%<br>
trial 3 = 211s; training accuracy = 96.74%; test accuracy = 95.95%<br><br>

SGD Alpha Comparisons:<br>
SGD, batch size = 64, alpha = .001:<br>
trial 1 = 225s; training accuracy = 33.79%; test accuracy = 13.62%<br>
SGD, batch size = 64, alpha = .1:<br>
trial 1 = 207s; training accuracy = 9.87%; test accuracy = 9.80%<br>
SGD, batch size = 64, alpha =.03:<br>
trial 1 = 210s; training accuracy = 97.72%; test accuracy = 98.16%<br>
trial 2 = 212s; training accuracy = 97.93%; test accuracy = 98.10%<br>
trial 3 = 224s; training accuracy = 97.82%; test accuracy = 98.09%<br>
SGD, batch size = 64, alpha = .05:<br>
trial 1 = 208s; training accuracy = 98.00%; test accuracy = 97.90%<br>
trial 2 = 212s; training accuracy = 97.62%; test accuracy = 97.97%<br>
trial 3 = 210s; training accuracy = 98.15%; test accuracy = 98.25%<br><br>

SGD batch size comparison:<br>
SGD, batch size = 32, alpha = .03:<br>
trial 1 = 231s; training accuracy = 98.29%; test accuracy = 98.21%<br>
trial 2 = 231s; training accuracy = 98.00%; test accuracy = 98.16%<br>
trial 3 = 230s; training accuracy = 98.13%; test accuracy = 98.59%<br>
SGD, batch size = 16, alpha = .03:<br>
trial 1 = 278s; training accuracy = 98.49%; test accuracy = 98.27%<br>
trial 2 = 279s; training accuracy = 98.48%; test accuracy = 98.72%<br>
trial 3 = 275s; training accuracy = 98.54%; test accuracy = 98.17%<br><br>

Final:<br>
SGD, batch size = 16, alpha = .03, loss = categorical cross entropy, epochs = 3<br>

In [89]:
from keras.optimizers.optimizer_v2.adam import Adam
# load cifar-10 dataset
from keras.datasets import cifar10
(x_cifar_train, y_cifar_train), (x_cifar_test, y_cifar_test) = cifar10.load_data()

print(x_cifar_train.shape)
print(x_cifar_test.shape)

print(y_cifar_train.shape)
print(y_cifar_test.shape)

# reshape data
x_cifar_train = x_cifar_train.reshape(50000,32,32,3)
x_cifar_test = x_cifar_test.reshape(10000,32,32,3)

# one-hot encode data
y_cifar_train = to_categorical(y_cifar_train)
y_cifar_test = to_categorical(y_cifar_test)

# create LeNet model
LeNet = Sequential()

LeNet.add(Conv2D(6, kernel_size=5, activation='relu', input_shape=(32,32,3)))
LeNet.add(MaxPooling2D(pool_size=(2, 2), strides=(2,2)))
LeNet.add(Conv2D(16, kernel_size=5, activation='relu'))
LeNet.add(MaxPooling2D(pool_size=(2, 2), strides=(2,2)))
LeNet.add(Conv2D(120, kernel_size=5, activation='relu'))
LeNet.add(Flatten())
LeNet.add(Dense(84, activation='relu'))
LeNet.add(Dense(10, activation='softmax'))

# compile model using accuracy to measure model performance
optimizer = keras.optimizers.Adam(lr=.0001)

LeNet.compile(
      optimizer=optimizer, 
      loss='categorical_crossentropy',
      metrics=['accuracy'])

LeNet.summary()

(50000, 32, 32, 3)
(10000, 32, 32, 3)
(50000, 1)
(10000, 1)
Model: "sequential_33"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_253 (Conv2D)         (None, 28, 28, 6)         456       
                                                                 
 max_pooling2d_22 (MaxPoolin  (None, 14, 14, 6)        0         
 g2D)                                                            
                                                                 
 conv2d_254 (Conv2D)         (None, 10, 10, 16)        2416      
                                                                 
 max_pooling2d_23 (MaxPoolin  (None, 5, 5, 16)         0         
 g2D)                                                            
                                                                 
 conv2d_255 (Conv2D)         (None, 1, 1, 120)         48120     
                                                           

In [90]:
LeNet.fit(
    x_cifar_train, 
    y_cifar_train, 
    batch_size = 256,
    validation_data=(x_cifar_test, y_cifar_test), epochs=25)

Epoch 1/25
196/196 [==============================] - 2s 8ms/step - loss: 3.7228 - accuracy: 0.1824 - val_loss: 2.3382 - val_accuracy: 0.2346
Epoch 2/25
196/196 [==============================] - 1s 6ms/step - loss: 2.1513 - accuracy: 0.2739 - val_loss: 1.9922 - val_accuracy: 0.3104
Epoch 3/25
196/196 [==============================] - 1s 6ms/step - loss: 1.9056 - accuracy: 0.3357 - val_loss: 1.8289 - val_accuracy: 0.3610
Epoch 4/25
196/196 [==============================] - 1s 6ms/step - loss: 1.7663 - accuracy: 0.3762 - val_loss: 1.7212 - val_accuracy: 0.3954
Epoch 5/25
196/196 [==============================] - 1s 6ms/step - loss: 1.6750 - accuracy: 0.4057 - val_loss: 1.6600 - val_accuracy: 0.4126
Epoch 6/25
196/196 [==============================] - 1s 6ms/step - loss: 1.6063 - accuracy: 0.4289 - val_loss: 1.6105 - val_accuracy: 0.4274
Epoch 7/25
196/196 [==============================] - 1s 6ms/step - loss: 1.5503 - accuracy: 0.4471 - val_loss: 1.5672 - val_accuracy: 0.4392
Epoch 

In [ ]:
# Evaluate
score = LeNet.evaluate(x_cifar_test, y_cifar_test)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

313/313 [==============================] - 1s 3ms/step - loss: 1.5756 - accuracy: 0.5650
Test loss: 1.5756491422653198
Test accuracy: 0.5649999976158142


LeNet Notes:<br><br>

learning rate testing (batch size = 64; epochs = 25; optimizer = adam):<br>
learning rate = .001(default):<br>
trial 1 = 102s; training accuracy = 77.50%; testing accuracy = 56.39%<br>
learning rate = .0001:<br>
trial 1 = 102s; training accuracy = 68.87%; testing accuracy = 56.70%<br>
learning rate = .01:<br>
trial 1 = 99s; training accuracy = 22.02%; testing accuracy = 18.47%<br>
learning rate = .005:<br>
trial 1 = 97s; training accuracy = 9.87%; testing accuracy = 10.00%<br><br>

batch size test (learning rate = .001; epochs = 25; optimizer = adam):<br>
Batch size 16:<br>
trial 1 = 298s; training accuracy = 65.78%; testing accuracy = 53.33%<br>
batch size 8:<br>
trial 1 = 289s; training accuracy = 61.46% testing accuracy = 49.78%<br>
batch size 128:<br>
trial 1 = 52s; training accuracy = 76.14%; testing accuracy = 53.09%<br>
trial 2 = 51s; training accuracy = 75.45%; testing accuracy = 54.67%<br>
batch size 256:<br>
trial 1 = 32s; training accuracy = 74.70%; testing accuracy = 53.83%<br>
batch size 256, learning rate .0001:<br>
trial 1 = 31s; training accuracy = 60.32%; testing accuracy = 53.82%<br>

In [92]:
# create feed forward approximation of LeNet
# assumption is that only the convolution layers are represented
# per instruction, "number of neurons = number of convolution filters"

feed_forward = keras.models.Sequential([
    keras.layers.Flatten(input_shape=(50000, 32, 32, 3)),
    keras.layers.Dense(6, activation='relu'),
    keras.layers.Dense(16, activation='relu'),
    keras.layers.Dense(120, activation='relu'),
    keras.layers.Dense(10, activation='softmax')
])

# compile model using accuracy to measure model performance
feed_forward.compile(
      optimizer='adam', 
      loss='categorical_crossentropy',
      metrics=['accuracy'])

feed_forward.summary()

ResourceExhaustedError: ignored

In [ ]:
feed_forward.fit(
    x_cifar_train, 
    y_cifar_train, 
    batch_size = 64,
    validation_data=(x_cifar_test, y_cifar_test), epochs=25)

In [ ]:
# Evaluate
score = feed_forward.evaluate(x_cifar_test, y_cifar_test)
print('Test loss:', score[0])
print('Test accuracy:', score[1])